server/src/main/java/org/opensearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java

```diff
@@ -93,6 +95,10 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
             // if its already a NO decision looking at the node, or we aren't configured to look at the host, return the decision
             return decision;
         }
+        if (INDEX_AUTO_EXPAND_REPLICAS_SETTING.get(allocation.metadata().getIndexSafe(shardRouting.index()).getSettings())
```

Comment:
Can you explain why you need to assign multiple replicas of same shard on same host? Today there's no use case where multiple replicas of same shard can be assigned on same node so this change may lead to unexpected issues. Have you tested multiple assignments on the node thoroughly? 
Also, if I have to guess then the change seems to target some non-critical indices where you need to ignore the awareness check but then if this is the case then why not reduce the replica count in the first place?

Reply:
agree ^^ but the change is not about need of assigning multiple replicas of same shard on same host, but the necessity of having the behavior for two deciders when shard copies coverage of nodes is 100%.   
Basically, two deciders aim to improve resilience when shard copies are limited. However, when shard copies are not limited, this behavior is unnecessary, so this commit disables them in those situations.

Reply:
Thanks for the replying @amberzsy ! See deciders are helping in achieving resiliency by making sure that replicas are spread out across nodes. If they were not and say all replicas were colocated then having a certain replica count would defeat the purpose. Therefore, deciders and replicas together are helping in achieving the desired resiliency. Disabling deciders conditionally and letting the shard assigned on same node vs keeping the shard unassigned isn't going to make any difference in terms of resiliency. However, this will unnecessary end up taking disk space, compute and other resources without providing any benefit whatsoever. And as I mentioned earlier, this may lead to unexpected issues as there may be assumptions in flows where no two shard copies can exist on the same node.

Reply:
i c. i think your point is if we bypass/disable the sameHostAllocation decider it potentially can end up with all shards (same number of datanodes) being accommodated to same host, which against the resiliency purpose.  correct?

Reply:
Yes, one of the many problems.

Reply:
sure. let me remove the skip logic in sameShardAllocation.